### PR TITLE
[RFC] Send LOGICAL_ERRORs to sentry

### PR DESCRIPTION
**Pull request by azat**
**Thoughts on this one?**

This is like an assert() in the code, so it is useful to know about them
as well.

This is not enabled by default, to turn it ON, set `send_crash_reports.send_logical_errors=true` in config

### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

### CODE REVIEW

**Comment by robot-ch-test-poll3** [comment]
*This is an automated comment for commit 2652ab36a26bec5c560ffb88470af0d62745608e with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/61997/2652ab36a26bec5c560ffb88470af0d62745608e/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

**Comment by nikitamikhaylov** [comment]
Ok, but let it be configurable at least https://clickhouse.com/docs/en/operations/server-configuration-parameters/settings#send_crash_reports

**Comment by azat** [comment]
Make sense - done. Configurable with `send_crash_reports.send_logical_errors`

**Comment by azat** [comment]
>Stateless tests (asan) [1/4] — Tests are not finished, fail: 2, passed: 564, skipped: 8
>Stateless tests (ubsan) [1/2] — Tests are not finished, fail: 4, passed: 1145, skipped: 14

- https://github.com/ClickHouse/ClickHouse/issues/62062

**Comment by azat** [comment]
Can someone take a look please?

**Comment by antaljanosbenjamin** [comment]
I can take a look at this today.

File: src/Daemon/SentryWriter.cpp [hunk 1]
```diff
@@ -222,8 +237,11 @@ void SentryWriter::onFault(int sig, const std::string & error_message, const Sta
...
 
-void SentryWriter::initialize(Poco::Util::LayeredConfiguration &) {}
-void SentryWriter::shutdown() {}
-void SentryWriter::onFault(int, const std::string &, const StackTrace &) {}
+void SentryWriter::initializeInstance(Poco::Util::LayeredConfiguration &) {}
+SentryWriter * SentryWriter::getInstance() { return nullptr; }
+
+SentryWriter::SentryWriter(Poco::Util::LayeredConfiguration &) {}
+SentryWriter::~SentryWriter() = default;
+void SentryWriter::onFault(int, const std::string &, const FramePointers &, size_t, size_t) {}
```
Comment (antaljanosbenjamin) [inline src/Daemon/SentryWriter.cpp:245]:
```suggestion
```

File: src/Daemon/SentryWriter.cpp [hunk 2]
```diff
@@ -133,35 +141,43 @@ void SentryWriter::initialize(Poco::Util::LayeredConfiguration & config)
...
+            sentry_set_tag("signal", strsignal(sig)); // NOLINT(concurrency-mt-unsafe) // not thread-safe but ok in this context
+            sentry_set_extra("signal_number", sentry_value_new_int32(sig));
+        }
+        else
+        {
+            /// Can be only LOGICAL_ERROR, but just in case.
+            int code = -sig_or_error;
+            sentry_set_tag("exception", DB::ErrorCodes::getName(code).data());
+            sentry_set_extra("exception_code", sentry_value_new_int32(code));
+        }
```
Comment (antaljanosbenjamin) [inline src/Daemon/SentryWriter.cpp:168]: I think it is better to separate not just by the sign of the things. There are two possibilities in my opinion:
1. keep thet `onFault` for signals and make another function for exceptions
2. introduce an enum/boolean flag to `onFault` that indicates whether it is signal or error

File: src/Daemon/SentryWriter.cpp [hunk 3]
```diff
@@ -133,35 +141,43 @@ void SentryWriter::initialize(Poco::Util::LayeredConfiguration & config)
...
 
-void SentryWriter::shutdown()
+SentryWriter::~SentryWriter()
 {
     if (initialized)
         sentry_shutdown();
 }
 
-void SentryWriter::onFault(int sig, const std::string & error_message, const StackTrace & stack_trace)
+void SentryWriter::onFault(int sig_or_error, const std::string & error_message, const FramePointers & frame_pointers, size_t offset, size_t size)
```
Comment (antaljanosbenjamin) [inline src/Daemon/SentryWriter.cpp:150]: I couldn't find definite answer in sentry docs about thread-safety. [They](https://docs.sentry.io/platforms/native/performance/instrumentation/custom-instrumentation/#thread-safety) mentioned that if something is not thread safe, then it is mentioned in the function's docs. `sentry_value_new_message_event` doesn't mention that, so my assumption is it is thread-safe. If I am right, then it is fine, if not, then we should make sure two logical errors happening at the same time won't cause any issues.

Previously we only reported crashes, so I am not sure if anybody thought about thread-safety, therefore I think we should make sure it is thread safe.

File: src/Daemon/SentryWriter.h [hunk 4]
```diff
@@ -14,14 +14,32 @@ class StackTrace;
 ///
 /// It is possible to send those reports to your own sentry account or account of consulting company you hired
 /// by overriding "send_crash_reports.endpoint" setting. "send_crash_reports.debug" setting will allow to do that for
-namespace SentryWriter
+class SentryWriter
```
Comment (antaljanosbenjamin) [inline src/Daemon/SentryWriter.h:17]: As I am trying to be cautious about what data I share with companies, I think it is a must to mention that this class from now on will also send some exceptions too. Could you please extend its documentation briefly?

**Review by antaljanosbenjamin — CHANGES_REQUESTED** [review]
The PR in general is good in my opinion. There are a few things to change, but nothing serious.

The only question I have remaining is have you tried it out at least once? If not, then I can make an attempt to see at least one `LOGICAL_ERROR` sent to sentry.

File: src/Daemon/SentryWriter.cpp (hunk 1 above)
Comment (azat) [reply src/Daemon/SentryWriter.cpp:245]: This placeholders required when sentry is disabled.

File: src/Daemon/SentryWriter.h (hunk 4 above)
Comment (azat) [reply src/Daemon/SentryWriter.h:17]: Documentation for the class you mean?

Documentation itself is already updated:

```
`send_logical_errors` – `LOGICAL_ERROR` is like an `assert`, it is a bug in ClickHouse. This boolean flag enables sending this exceptions to sentry (default: `false`).
```

File: src/Daemon/SentryWriter.cpp (hunk 3 above)
Comment (azat) [reply src/Daemon/SentryWriter.cpp:150]: Yes, it looks safe, global variables guarded with mutex (thought 100% it is hard to tell of course).

File: src/Daemon/SentryWriter.cpp (hunk 2 above)
Comment (azat) [reply src/Daemon/SentryWriter.cpp]: Done.

File: src/Daemon/SentryWriter.h (hunk 4 above)
Comment (azat) [reply src/Daemon/SentryWriter.h:18]: Updated brief comment for class as well.

File: src/Daemon/SentryWriter.cpp (hunk 1 above)
Comment (azat) [reply src/Daemon/SentryWriter.cpp]: Here is an example - https://s3.amazonaws.com/clickhouse-test-reports/61997/8b39cb2c6d08a2e86867e068bef0a8689362634a/fast_test/run.log

```
2024-04-11 17:21:08 /ClickHouse/src/Daemon/SentryWriter.cpp:260:20: error: out-of-line definition of 'onFault' does not match any declaration in 'SentryWriter'
2024-04-11 17:21:08   260 | void SentryWriter::onFault(int, const std::string &, const FramePointers &, size_t, size_t) {}
2024-04-11 17:21:08       |                    ^~~~~~~

```

Comment (antaljanosbenjamin) [reply src/Daemon/SentryWriter.cpp]: Oh, I totally missed this, you are right.

File: src/Daemon/SentryWriter.h (hunk 4 above)
Comment (antaljanosbenjamin) [reply src/Daemon/SentryWriter.h:18]: > Documentation for the class you mean?

Yes, I meant that, thanks for updating it.

**Review by antaljanosbenjamin — APPROVED** [review]

**Comment by azat** [comment]
@antaljanosbenjamin does it waits something else?

**Comment by antaljanosbenjamin** [comment]
Sorry, I forgot to merge it. Merging now.

**Pull request merged**